### PR TITLE
only players near you can see text chat and hear voice

### PR DESCRIPTION
**Pull request by plally**

### CODE REVIEW

File: gamemode/modules/chat/sv_chat.lua [hunk 1]
```diff
@@ -0,0 +1,13 @@
...
+        return false
+    end
+    
+    if listenerPos:Distance( speakerPos ) > GAMEMODE.Config.Chat.CHAT_DISTANCE then
+        return false
+    end
+
+    return true
+end
+
```
Comment (plally) [inline gamemode/modules/chat/sv_chat.lua:17]:
```suggestion
```

File: gamemode/modules/chat/sv_chat.lua [hunk 2]
```diff
@@ -0,0 +1,27 @@
+function RVR.canHear( listener, speaker )
```
Comment (samuelWilliams99) [inline gamemode/modules/chat/sv_chat.lua]: Can we put this in like `RVR.Chat`
You'd also need
```lua
RVR.Chat = RVR.Chat or {}
local Chat = RVR.Chat
```
To fit in with other modules

File: gamemode/modules/chat/sv_chat.lua [hunk 3]
```diff
@@ -0,0 +1,27 @@
...
+    if listenerPos:Distance( speakerPos ) > GAMEMODE.Config.Chat.CHAT_DISTANCE then
+        return false
+    end
+    
+    return true
+end
+
+function GM:PlayerCanSeePlayersChat( text, teamOnly, listener, speaker )
+    if teamOnly then
+        -- TODO check party system?
```
Comment (samuelWilliams99) [inline gamemode/modules/chat/sv_chat.lua]: Can use a hook on `PlayerCanSeePlayersChat` to detect this, so this TODO isn't needed

File: gamemode/modules/chat/sv_chat.lua [hunk 4]
```diff
@@ -0,0 +1,27 @@
+function RVR.canHear( listener, speaker )
+    if not listener:Alive() or not speaker:Alive() then
+        return false
+    end
+    
+    local speakerPos = speaker:GetPos()
+    local listenerPos = listener:GetPos()
+
+    if listenerPos:Distance( speakerPos ) > GAMEMODE.Config.Chat.CHAT_DISTANCE then
```
Comment (samuelWilliams99) [inline gamemode/modules/chat/sv_chat.lua]: Idk how much this matters, but I often see people using `DistToSqr` over `Distance` and squaring the config value, as apparently that's faster. Up to you as to if it's worth the slightly less readable code.

**Review by samuelWilliams99 — CHANGES_REQUESTED** [review]
Only really want the function moved into a table

File: gamemode/modules/chat/sv_chat.lua (hunk 4 above)
Comment (plally) [reply gamemode/modules/chat/sv_chat.lua]: ~~i dont think efficiency will be an issue here as this is only ran once on chat~~ nvm this is used in voice as well, not sure how often that is ran

**Review by samuelWilliams99 — APPROVED** [review]

**Review by ghost — APPROVED** [review]
Looks good to me

**Pull request merged**